src/degree_two_reduction.jl

```diff
@@ -300,11 +325,38 @@ function find_degree2_chains(
         end
         chain_path =
             _get_complete_chain(adj_matrix, node, reduced_indices, irreducible_indices)
-        if adj_matrix[chain_path[1], chain_path[end]] != 0
-            @warn "Nodes $(chain_path[1]) and $(chain_path[end]) already have a parallel path, skipping chain creation."
-        else
-            arc_map[chain_path[1], chain_path[end]] = chain_path
+        valid_chain_path = _find_longest_valid_chain(adj_matrix, chain_path)
+        if !isempty(valid_chain_path)
+            arc_map[valid_chain_path[1], valid_chain_path[end]] = valid_chain_path
         end
     end
     return arc_map
 end
+
+function _find_longest_valid_chain(
+    adj_matrix::SparseArrays.SparseMatrixCSC,
+    chain_path::Vector{Int},
+)
+    if adj_matrix[chain_path[1], chain_path[end]] == 0
+        return chain_path
+    else
+        @warn "Nodes $(chain_path[1]) and $(chain_path[end]) already have a parallel path, searching for valid subchains."
+        subchains = Vector{Int}[]
+        n = length(chain_path)
+        for i in 1:n
+            for j in i:n
+                push!(subchains, chain_path[i:j])
+            end
+        end
+        sort!(subchains; by = length, rev = true)
+        filter!(x -> length(x) > 2, subchains)
```

Comment:
why this changes? doing a sort! and a filter with an anonymous function in a hot loop isn't a good idea. We need to discuss this one

Reply:
something like this would be internally optimized by Julia for not huge chains 
```
sort([x for x in data if length(x) > 2], by = length, rev = true)
```